org.eclipse.xtext.ui/src/org/eclipse/xtext/ui/editor/model/edit/CompositeModificationWrapper.java

```diff
@@ -0,0 +1,146 @@
+ * @noreference This class is not intended to be referenced by clients.
+ * 
+ * @author Moritz Eysholdt - Initial contribution and API
+ */
+public class CompositeModificationWrapper implements IModification, BatchModification.IBatchableModification {
+
+	public static class ModificationPart<T extends Notifier> implements IChangeSerializer.IModification<T> {
+		private final T element;
+		private final ICompositeModificationContext.IModificationPart<T> part;
+
+		public ModificationPart(IModificationPart<T> part, T element) {
+			super();
+			this.part = part;
+			this.element = element;
+		}
+
+		public void apply() {
+			part.apply(element);
+		}
+
+		@Override
+		public void modify(T context) {
+			part.apply(context);
+		}
+
+	}
+
+	protected class MultiModificationContext implements ICompositeModificationContext<EObject> {
+		private final EObject element;
+		private final IChangeSerializer serializer;
+
+		public MultiModificationContext(EObject element, IChangeSerializer serializer) {
+			this.element = element;
+			this.serializer = serializer;
+		}
+
+		@Override
+		public <N extends Notifier> void addModification(N toModify, IModificationPart<N> modification) {
+			serializer.addModification(toModify, new ModificationPart<N>(modification, toModify));
+		}
+
+		@Override
+		public EObject getElement() {
+			return this.element;
+		}
+
+		@Override
+		public Issue getIssue() {
+			return issue;
+		}
+
+		@Override
+		public ITextRegionDiffBuilder getModifyableDocument(XtextResource resource) {
+			return serializer.getModifyableDocument(resource);
+		}
+
+		@Override
+		public boolean isUpdateCrossReferences() {
+			return updateCrossReferences;
+		}
+
+		@Override
+		public boolean isUpdateRelatedFiles() {
+			return updateRelatedFiles;
+		}
+
+		@Override
+		public void setUpdateCrossReferences(boolean value) {
+			updateCrossReferences = value;
+		}
+
+		@Override
+		public void setUpdateRelatedFiles(boolean value) {
+			updateRelatedFiles = value;
+		}
+	}
+
+	private final Issue issue;
+	private final ICompositeModification<EObject> modification;
+	private boolean updateCrossReferences = true;
+	private boolean updateRelatedFiles = true;
+
+	@SuppressWarnings("unchecked")
+	public CompositeModificationWrapper(Issue issue, ICompositeModification<?> modification) {
+		this.issue = issue;
+		this.modification = (ICompositeModification<EObject>) modification;
+	}
+
+	@Override
+	public void apply(EObject object, IChangeSerializer serializer) {
+		modification.apply(object, new MultiModificationContext(object, serializer));
+	}
+
+	@Override
+	public void apply(IModificationContext ctx) throws Exception {
+		IXtextDocument document = ctx.getXtextDocument();
+		BatchModification batch = document.readOnly(r -> r.getResourceServiceProvider().get(BatchModification.class));
+		batch.setDocument(document);
+		batch.apply(Collections.singleton(this), new NullProgressMonitor());
```

Comment:
Any chance to funnel a cancel indicator / progress monitor through this API?

Reply:
to be discussed / fixed in https://github.com/eclipse/xtext-eclipse/issues/420